_posts/2025-04-23-#31444.md

```diff
@@ -10,17 +10,95 @@ status: upcoming
 commit:
 ---
 
-_Notes and questions to follow soon!_
-
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
 
+### Motivations and Background
+- Miners use a mempool to store and sort unconfirmed transactions that they can include in blocks to collect their fees.
+  Block building (done by the
+[`BlockAssembler`](https://github.com/bitcoin/bitcoin/blob/cdc32994feadf3f15df3cfac5baae36b4b011462/src/node/miner.h#L145))
+is the process of selecting ~4MWu worth of mempool transactions while trying to maximize the total fees.
+- Non-mining nodes might never use the block builder, but are still interested in comparing the incentive compatibility
+  of transactions as a metric for whether it is worth keeping: if the mempool exceeds its maximum memory allowance, it
+should evict the transactions that are least likely to be mined soon.
+- The [block building algorithm](https://github.com/bitcoin/bitcoin/blob/cdc32994feadf3f15df3cfac5baae36b4b011462/src/node/miner.cpp#L296)
+on master uses ancestor set-based sort to dynamically linearize mempool transactions by including ancestor sets in order of
+highest **ancestor score** (minimum between ancestor feerate and individual feerate). As transactions are included,
+their descendants' ancestor scores can change; the `BlockAssembler` keeps a copy of mempool entries with their updated
+ancestor information in a separate map, which avoids modifying the mempool itself. We covered this algorithm in more
+detail in a [previous meeting](/26152).
+- In effect, ancestor set-based sorting produces a linearization of mempool transactions (from most to least likely
+  to be mined), though they have not been called as such.
+- Notably, the only way to determine the linearization position of a mempool transaction this way is to run the
+  block building algorithm until this transaction is selected. This means that finding the "least likely to be
+mined" transaction (e.g. for eviction) requires building an entire mempool worth of block templates.
+- Instead, eviction runs a similar algorithm but with ascending **descendant score** (maximum between descendant
+  feerate and individual feerate). This also linearizes transactions in an approximate order of least to most likely
+to be mined, but the linearization is not exactly opposite to the result of ancestor set-based sorting. This asymmetry
+is problematic:
+    - What if the first transaction selected for eviction is also the transaction that would be selected first for
+      mining? This cache eviction strategy is throwing away the most useful piece of data.
```

Comment:
> This cache eviction strategy is throwing away the most useful piece of data.

"cache" to refer to mempool might be unnecessarily confusing, but I think this whole second sentence could also just be left out?

Reply:
removed